### DISCUSSION TITLE
How do I write an asset that's computed using a k8s/kubernetes job?

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#14264] [Q&A] [answered]
I want to define an asset that, whenever I click "Materialize" on it or it's automatically materialized, launches a particular k8s job.  I want to be able to customize the k8s job for this particular asset, so I don't want to deal with using the kubernetes executor.  I want something that functions similarly to Airflow's `KubernetesPodOperator`.

**Comment by sryza** [comment, marked as answer]
Here's an example of how to make a single asset that's computed by executing a k8s job:

```python
from dagster import asset
from dagster_k8s import execute_k8s_job


@asset
def table1(context) -> None:
    execute_k8s_job(
        context,
        image="YOUR_IMAGE_HERE",
        command=["/bin/sh", "-c"],
        args=["update_table_1.sh"],
    )
```

If you want this asset to depend on other assets, you can use `non_argument_deps` to specify the dependency.

Here's a pattern that's useful if you want to generate a set of similar asset assets that are all computed using a k8s job:
```python
from dagster import asset
from dagster_k8s import execute_k8s_job

def make_k8s_asset(name):
    """
    Returns a new asset definition with the given name.
    
    When the asset is told to materialize, a k8s job will execute that invokes a script with the asset's name.
    """

    @asset(name=name)
    def k8s_asset(context) -> None:
        execute_k8s_job(
            context,
            image="YOUR_IMAGE_HERE",
            command=["/bin/sh", "-c"],
            args=[f"update_{name}.sh"],
        )

    return k8s_asset


table_a = make_k8s_asset("table_a")
table_b = make_k8s_asset("table_b")
```

**Reply by gordondavies** [reply]
Thanks for this @sryza, its not obvious what the container is returning here, is the execute_k8s_job collecting stdout from the pod, or is there some other mechanism to capture the output (e.g. 'table_a' in this case). Or am I misunderstanding the concept of an asset here, and in fact the container is just building 'table_a' somewhere like a regular 'op' might?

**Reply by sryza** [reply]
@gibsondan -  do you know the answer to the above question about collecting stdout from the pod launched by `execute_k8s_job`?

**Reply by gibsondan** [reply]
Hi @gordondavies - in this particular case the container doesn't return anything. make_k8s_asset is returning a Dagster asset definition, not a particular value. If you're using execute_k8s_job to execute arbitrary k8s containers, Dagster doesn't provide any particular mechanism for capturing the output or result of that k8s job.

**Reply by gibsondan** [reply]
'make_k8s_asset' here is a Dagster asset factory - when it's called, it returns a Dagster asset definition, so `table_a` is more like a function that runs the kubernetes job than it is a particular output or result.

**Reply by gordondavies** [reply]
Ok, thanks! So, if the container does not return anything to Dagster, and Dagster has no mechanism for capturing the output of container's execution, what's the benefit of using an asset here instead of an op?

EDIT: Ok - I realise now - its so that the k8s op is automatically executed when asset is 'materialised'. 

Thanks again!

**Reply by sryza** [reply]
> EDIT: Ok - I realise now - its so that the k8s op is automatically executed when asset is 'materialised'.

Exactly
